pkg/platform/runtime/runtime.go

```diff
@@ -66,19 +65,43 @@ func New(target setup.Targeter) (*Runtime, error) {
 	return r, err
 }
 
-func (r *Runtime) Update(msgHandler setup.MessageHandler) error {
+// Update updates the runtime by downloading all necessary artifacts from the Platform and installing them locally.
+// This function is usually called, after New() returned with a NeedsUpdateError
+func (r *Runtime) Update(msgHandler *runbits.RuntimeMessageHandler) error {
 	logging.Debug("Updating %s#%s @ %s", r.target.Name(), r.target.CommitUUID(), r.target.Dir())
-	if err := setup.New(r.target, msgHandler).Update(); err != nil {
-		return errs.Wrap(err, "Update failed")
-	}
-	rt, err := new(r.target)
+
+	// Run the setup function (the one that produces runtime events) in the background...
+	prod := events.NewRuntimeEventProducer()
+	var setupErr error
+	go func() {
+		defer prod.Close()
+
+		if err := setup.New(r.target, prod).Update(); err != nil {
+			setupErr = errs.Wrap(err, "Update failed")
+			return
+		}
+		rt, err := new(r.target)
+		if err != nil {
+			setupErr = errs.Wrap(err, "Could not reinitialize runtime after update")
+			return
+		}
+		*r = *rt
+	}()
+
+	// ... and handle the runtime events in the main thread
+	err := msgHandler.HandleEvents(prod.Events())
```

Comment:
Is `HandleEvents` blocking? Would be good if we convey that in the method name, or have an additional `msgHandler.Wait()`.

Reply:
Yes, I was thinking about that name for a long time, but wasn't really clear on what needed to be conveyed.

I like `WaitForAllEvents()`, `HandleAllEvents` or `HandleAllEventsAndWait()`.

The second would be my favorite.

Reply:
I don't think `HandleAllEvents` conveys that it blocks. Out of these options I would favor `WaitForAllEvents()`.